verilog/tools/ls/symbol-table-handler.cc

```diff
@@ -0,0 +1,249 @@
+      !status.ok()) {
+    // if failed to parse
+    LOG(WARNING) << "Failed to parse file list in " << projectpath << ":  "
+                 << status;
+    return;
+  }
+  // update include directories in project
+  for (const auto &incdir : filelist.preprocessing.include_dirs) {
+    LOG(INFO) << "Adding include path:  " << incdir;
+    currproject->AddIncludePath(incdir);
+  }
+  // add files from file list to the project
+  for (auto &incfile : filelist.file_paths) {
+    auto incsource = currproject->OpenIncludedFile(incfile);
+    if (!incsource.ok()) {
+      LOG(WARNING) << "File included in " << projectpath
+                   << " not found:  " << incfile << ":  " << incsource.status();
+      continue;
+    }
+    LOG(INFO) << "Creating symbol table for:  " << incfile;
+    BuildSymbolTableFor(*incsource.value());
+  }
+}
+
+bool IsStringViewContained(absl::string_view origin, absl::string_view substr) {
+  const int from = std::distance(origin.begin(), substr.begin());
+  const int to = std::distance(origin.begin(), substr.end());
+  if (from < 0) return false;
+  if (to > static_cast<int>(origin.length())) return false;
+  return true;
+}
+
+const SymbolTableNode *ScanReferenceComponents(
+    const ReferenceComponentNode *ref, absl::string_view symbol) {
+  if (IsStringViewContained(ref->Value().identifier, symbol))
+    return ref->Value().resolved_symbol;
+  for (const auto &childref : ref->Children()) {
+    const SymbolTableNode *resolved =
+        ScanReferenceComponents(&childref, symbol);
+    if (resolved) return resolved;
+  }
+  return nullptr;
+}
+
+const SymbolTableNode *SymbolTableHandler::ScanSymbolTreeForDefinition(
+    const SymbolTableNode *context, absl::string_view symbol) {
+  if (!context) {
+    return nullptr;
+  }
+  // TODO (glatosinski): reduce searched scope by utilizing information from
+  // syntax tree?
+  for (const auto &ref : context->Value().local_references_to_bind) {
+    if (ref.Empty()) continue;
+    const SymbolTableNode *resolved =
+        ScanReferenceComponents(ref.components.get(), symbol);
+    if (resolved) return resolved;
+  }
+  for (const auto &child : context->Children()) {
+    auto res = ScanSymbolTreeForDefinition(&child.second, symbol);
+    if (res) {
+      return res;
+    }
+  }
+  return nullptr;
+}
+
+std::vector<verible::lsp::Location> SymbolTableHandler::FindDefinition(
+    const verible::lsp::DefinitionParams &params,
+    const verilog::BufferTrackerContainer &parsed_buffers) {
+  const absl::Time finddefinition_start = absl::Now();
+  if (files_dirty_) {
+    BuildProjectSymbolTable();
+  }
+  absl::string_view filepath = LSPUriToPath(params.textDocument.uri);
+  if (filepath.empty()) {
+    LOG(ERROR) << "Could not convert URI " << params.textDocument.uri
+               << " to filesystem path." << std::endl;
+    LOG(INFO) << "textDocument/definition processing time:  "
+              << absl::ToInt64Milliseconds(absl::Now() - finddefinition_start)
```

Comment:
The `absl::Duration` type can already print on a stream, so you don't have to wrap the whole thing in `absl::ToInt64Milliseconds()`.

```c++
LOG(...) << (absl::Now() - finddefinition_start);
```

would be sufficient. Advantage: it will automatically add the unit (e.g. `ms`), and if the time is short, say a few microseconds, it will actually print it as microseconds. So easier to read, and less work writing the log message.